File: programmers/kakao2022/disappearingplatform/DisappearingPlatform.py
# Delta 값: 상, 하, 좌, 우 이동 (행, 열)
DELTA_ROW = [-1, 1, 0, 0]
DELTA_COL = [0, 0, -1, 1]

# 보드 경계 확인 함수
def is_out_of_bounds(row, col, board_height, board_width):
    return row < 0 or row >= board_height or col < 0 or col >= board_width

# 게임을 재귀적으로 진행하며 최적의 결과를 찾는 함수
def play_game(board, a_row, a_col, b_row, b_col, move_count, is_a_turn, board_height, board_width):
    current_row = a_row if is_a_turn else b_row
    current_col = a_col if is_a_turn else b_col

    # 1. 종료 조건
    # 1.1. 현재 위치가 경계를 벗어나거나 플랫폼이 없으면 패배
    if is_out_of_bounds(current_row, current_col, board_height, board_width) or board[current_row][current_col] == 0:
        return (False, move_count)

    # 1.2. 이동 가능한 위치가 없으면 패배
    can_move = False
    for i in range(4):
        next_row = current_row + DELTA_ROW[i]
        next_col = current_col + DELTA_COL[i]
        if not is_out_of_bounds(next_row, next_col, board_height, board_width) and board[next_row][next_col] == 1:
            can_move = True
            break
    if not can_move:
        return (False, move_count)

    # 2. 재귀 단계
    # 2.1. 현재 플레이어의 플랫폼 제거
    board[current_row][current_col] = 0

    # 2.2. 모든 가능한 이동에 대해 재귀 호출
    can_win = False
    min_moves_to_win = float('inf')
    max_moves_to_lose = 0

    for i in range(4):
        next_row = current_row + DELTA_ROW[i]
        next_col = current_col + DELTA_COL[i]

        if is_out_of_bounds(next_row, next_col, board_height, board_width) or board[next_row][next_col] == 0:
            continue

        if is_a_turn:
            next_result = play_game(board, next_row, next_col, b_row, b_col, move_count + 1, False, board_height, board_width)
        else:
            next_result = play_game(board, a_row, a_col, next_row, next_col, move_count + 1, True, board_height, board_width)

        if not next_result[0]:
            can_win = True
            min_moves_to_win = min(min_moves_to_win, next_result[1])
        else:
            max_moves_to_lose = max(max_moves_to_lose, next_result[1])

    # 2.3. 백트래킹: 플랫폼 복원
    board[current_row][current_col] = 1

    # 3. 결과 반환
    if can_win:
        return (True, min_moves_to_win)
    else:
        return (False, max_moves_to_lose)

# 최적의 이동 횟수를 계산하는 함수
def solution(board, aloc, bloc):
    board_height = len(board)
    board_width = len(board[0])

    result = play_game(board, aloc[0], aloc[1], bloc[0], bloc[1], 0, True, board_height, board_width)
    return result[1]

# 테스트 코드
if __name__ == "__main__":
    board1 = [[1, 1, 1], [1, 1, 1], [1, 1, 1]]
    aloc1 = [1, 0]
    bloc1 = [1, 2]
    print(f"Test Case 1: {solution(board1, aloc1, bloc1)} (Expected: 5)")

    board2 = [[1, 1, 1], [1, 0, 1], [1, 1, 1]]
    aloc2 = [1, 0]
    bloc2 = [1, 2]
    print(f"Test Case 2: {solution(board2, aloc2, bloc2)} (Expected: 4)")

    board3 = [[1, 1, 1, 1, 1]]
    aloc3 = [0, 0]
    bloc3 = [0, 4]
    print(f"Test Case 3: {solution(board3, aloc3, bloc3)} (Expected: 4)")

    board4 = [[1]]
    aloc4 = [0, 0]
    bloc4 = [0, 0]
    print(f"Test Case 4: {solution(board4, aloc4, bloc4)} (Expected: 0)")

File: programmers/kakao2022/disappearingplatform/DisappearingPlatform.js
// Delta 값: 상, 하, 좌, 우 이동 (행, 열)
const DELTA_ROW = [-1, 1, 0, 0];
const DELTA_COL = [0, 0, -1, 1];

// 보드 경계 확인 함수
function isOutOfBounds(row, col, boardHeight, boardWidth) {
    return row < 0 || row >= boardHeight || col < 0 || col >= boardWidth;
}

// 게임을 재귀적으로 진행하며 최적의 결과를 찾는 함수
function playGame(board, aRow, aCol, bRow, bCol, moveCount, isATurn, boardHeight, boardWidth) {
    let currentRow = isATurn ? aRow : bRow;
    let currentCol = isATurn ? aCol : bCol;

    // 1. 종료 조건
    // 1.1. 현재 위치가 경계를 벗어나거나 플랫폼이 없으면 패배
    if (isOutOfBounds(currentRow, currentCol, boardHeight, boardWidth) || board[currentRow][currentCol] === 0) {
        return { canWin: false, moveCount };
    }

    // 1.2. 이동 가능한 위치가 없으면 패배
    let canMove = false;
    for (let i = 0; i < 4; i++) {
        let nextRow = currentRow + DELTA_ROW[i];
        let nextCol = currentCol + DELTA_COL[i];
        if (!isOutOfBounds(nextRow, nextCol, boardHeight, boardWidth) && board[nextRow][nextCol] === 1) {
            canMove = true;
            break;
        }
    }
    if (!canMove) {
        return { canWin: false, moveCount };
    }

    // 2. 재귀 단계
    // 2.1. 현재 플레이어의 플랫폼 제거
    board[currentRow][currentCol] = 0;

    // 2.2. 모든 가능한 이동에 대해 재귀 호출
    let canWin = false;
    let minMovesToWin = Infinity;
    let maxMovesToLose = 0;

    for (let i = 0; i < 4; i++) {
        let nextRow = currentRow + DELTA_ROW[i];
        let nextCol = currentCol + DELTA_COL[i];

        if (isOutOfBounds(nextRow, nextCol, boardHeight, boardWidth) || board[nextRow][nextCol] === 0) {
            continue;
        }

        let nextResult;
        if (isATurn) {
            nextResult = playGame(board, nextRow, nextCol, bRow, bCol, moveCount + 1, false, boardHeight, boardWidth);
        } else {
            nextResult = playGame(board, aRow, aCol, nextRow, nextCol, moveCount + 1, true, boardHeight, boardWidth);
        }

        if (!nextResult.canWin) {
            canWin = true;
            minMovesToWin = Math.min(minMovesToWin, nextResult.moveCount);
        } else {
            maxMovesToLose = Math.max(maxMovesToLose, nextResult.moveCount);
        }
    }

    // 2.3. 백트래킹: 플랫폼 복원
    board[currentRow][currentCol] = 1;

    // 3. 결과 반환
    if (canWin) {
        return { canWin: true, moveCount: minMovesToWin };
    } else {
        return { canWin: false, moveCount: maxMovesToLose };
    }
}

// 최적의 이동 횟수를 계산하는 함수
function solution(board, aloc, bloc) {
    let boardHeight = board.length;
    let boardWidth = board[0].length;

    let result = playGame(board, aloc[0], aloc[1], bloc[0], bloc[1], 0, true, boardHeight, boardWidth);
    return result.moveCount;
}

// 테스트 코드
let board1 = [[1, 1, 1], [1, 1, 1], [1, 1, 1]];
let aloc1 = [1, 0];
let bloc1 = [1, 2];
console.log(`Test Case 1: ${solution(board1, aloc1, bloc1)} (Expected: 5)`);

let board2 = [[1, 1, 1], [1, 0, 1], [1, 1, 1]];
let aloc2 = [1, 0];
let bloc2 = [1, 2];
console.log(`Test Case 2: ${solution(board2, aloc2, bloc2)} (Expected: 4)`);

let board3 = [[1, 1, 1, 1, 1]];
let aloc3 = [0, 0];
let bloc3 = [0, 4];
console.log(`Test Case 3: ${solution(board3, aloc3, bloc3)} (Expected: 4)`);

let board4 = [[1]];
let aloc4 = [0, 0];
let bloc4 = [0, 0];
console.log(`Test Case 4: ${solution(board4, aloc4, bloc4)} (Expected: 0)`);